### DISCUSSION TITLE
Graceful Shutdown stops Jetty server

### DISCUSSION THREAD

**Discussion by ltagliamonte-dd** [dropwizard/dropwizard#4770] [General]
I'm trying to gracefully shutdown my app on TERM signal.
So far my code look like:
```
public class ServerApplication extends Application<ServerConfiguration> {
    public static void main(String[] args) throws Exception {
      
        ServerApplication application = new ServerApplication();
        application.run(args);

        Runtime.getRuntime().addShutdownHook(new Thread("SHUTDOWN_HOOK") {
            @Override
            public void run() {
                System.out.println("!!! application is shutting down !!!");
                application.delayShutdown();
                application.stop();
                System.out.println("<---------- Shutdown complete ---------->");
            }
        });
    }
```

the `delayShutdown` method will set a boolean to false and sleep for N seconds, i'm doing this in order to make the readiness endpoint of my app fail so that Kubernetes will stop sending traffic to my container.

The code get correctly executed but I've found out that as soon the TERM signal is received Jetty will close the socket and all in new requests will fail and I don't want this, I need for the server to be fully available until the sleep ends.

Running a while true loop with a test curl on TERM will get me:
```
curl: (7) Failed to connect to localhost port 80: Connection refused
curl: (7) Failed to connect to localhost port 80: Connection refused
curl: (7) Failed to connect to localhost port 80: Connection refused
curl: (7) Failed to connect to localhost port 80: Connection refused
```

I've tried shutdownGracePeriod with no luck:
```
server:
  shutdownGracePeriod: 90s
```

Would really appreciate the community help on this matter, Thank you!

**Comment by pstackle** [comment]
Hi @ltagliamonte-dd 
What you are describing is very similar to one of the features that the [dropwizard-health](https://github.com/dropwizard/dropwizard-health) library provides. 

It configures a delayed shutdown handler with a shutdown wait period. Once Jetty receives the signal to shutdown, the delayed shutdown handler (which is registered as the first handler) is triggered and it immediately sets the app's health status to false and then goes to sleep for the duration of the wait period. During this time, your app instance's health check would be returning an unhealthy status to calls to the `/health-check` endpoint. Because this is the first shutdown handler, then the rest of Jetty's shutdown process hasn't been triggered yet, so requests should continue to be accepted and processed. The wait period should be configured to be long enough to allow for your app instance to no longer be deemed healthy and to allow for it to finish processing any remaining requests. Once the wait period expires, the shutdown handler wakes up and returns, which allows Jetty to continue the rest of its shutdown process.

You can see the lines of code where the `HealthCheckBundle` registers the shutdown handler below.
https://github.com/dropwizard/dropwizard-health/blob/master/src/main/java/io/dropwizard/health/core/HealthCheckBundle.java#L79-L84

Does that help?

**Comment by ltagliamonte-dd** [comment]
Thank you @pstackle !!
Will look into this!! Really appreciated.

**Comment by harpreet86** [comment]
Hi @pstackle , Hi @joschi 

We added below configuration in our project:
```
health:
  delayedShutdownHandlerEnabled: true
  shutdownWaitPeriod: '60s'
```

On debugging we can see that DelayedShutdownHook is registered and called successfully on graceful termination marking application unhealthy but when hitting the application  /health-check URL it returns application status as healthy. Please suggest if we are missing something or is it some bug.

We are using dropwizard version 4.0.7
